File: p543.cpp
#include <bits/stdc++.h>
#include <TreeNode.hpp>

using namespace std;

class Solution {
public:
    map<TreeNode*, int> depths;
    int depth(TreeNode* root) {
        if(root == NULL) {
            return 0;
        } else if(depths[root]) {
            return depths[root];
        }
        int result = max(depth(root->left) + 1, depth(root->right) + 1);
        depths[root] = result;
        return result;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter = 0;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()) {
            if(!q.front()) {
                q.pop();
                continue;
            }
            TreeNode* curr = q.front();
            q.push(curr->left);
            q.push(curr->right);
            q.pop();
            diameter = max(diameter, depth(curr->left) + depth(curr->right));
        }
        return diameter;
    }
};

File: p543.py
from functools import lru_cache

from DS.TreeNode import TreeNode

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        if root is None:
            return 0

        @lru_cache(None)
        def height(node):
            if node is None:
                return -1
            return 1 + max(height(node.left), height(node.right))

        def dfs(node):
            if node is None:
                return -1
            return max(2 + height(node.left) + height(node.right), \
                       dfs(node.left), dfs(node.right))

        return dfs(root)
